src/stream/src/executor/temporal_filter.rs

```diff
@@ -0,0 +1,466 @@
+        ctx: ActorContextRef,
+        input: BoxedExecutor,
+        pk_indices: PkIndices,
+        executor_id: u64,
+        state_table: StateTable<S>,
+        time_col_idx: usize,
+        window_size: ScalarImpl,
+        chunk_size: usize,
+    ) -> Self {
+        let schema = input.schema().clone();
+        Self {
+            ctx,
+            input: Some(input),
+            pk_indices,
+            identity: format!("TemporalFilterExecutor {:X}", executor_id),
+            schema,
+            state_table,
+            time_col_idx,
+            window_size,
+            chunk_size,
+            buffer: BTreeMap::new(),
+        }
+    }
+
+    fn apply_batch(
+        &mut self,
+        data_chunk: &DataChunk,
+        ops: Vec<Op>,
+        lower_condition: &BoxedExpression,
+    ) -> (Vec<Op>, Bitmap) {
+        debug_assert_eq!(ops.len(), data_chunk.capacity());
+        let mut new_ops = Vec::with_capacity(ops.len());
+        let mut new_visibility = BitmapBuilder::with_capacity(ops.len());
+        let mut last_res = false;
+
+        let lower_results_array = lower_condition.eval_infallible(data_chunk, |err| {
+            self.ctx.on_compute_error(err, self.identity())
+        });
+        let lower_results = must_match!(
+            lower_results_array.deref(), ArrayImpl::Bool(eval_results) => eval_results);
+
+        for ((row_ref, op), lower_result) in data_chunk
+            .rows_with_holes()
+            .zip_eq(ops.into_iter())
+            .zip_eq(lower_results.to_bitmap().iter())
+        {
+            if let Some(row_ref) = row_ref {
+                let res = lower_result;
+                match op {
+                    Op::Insert | Op::Delete => {
+                        new_ops.push(op);
+                        new_visibility.append(res);
+                    }
+                    Op::UpdateDelete => {
+                        last_res = res;
+                    }
+                    Op::UpdateInsert => {
+                        if last_res == res {
+                            new_ops.push(Op::UpdateDelete);
+                            new_ops.push(Op::UpdateInsert);
+                        } else {
+                            new_ops.push(Op::Delete);
+                            new_ops.push(Op::Insert);
+                        }
+                        new_visibility.append(last_res);
+                        new_visibility.append(res);
+                    }
+                }
+                if lower_result {
+                    let row = row_ref.to_owned_row();
+                    // Null event time should not exist in the row since it passed the filter.
+                    let timestamp = row[self.time_col_idx].as_ref().unwrap().clone();
+                    let pk = row.by_indices(self.pk_indices());
+                    match op {
+                        Op::Insert | Op::UpdateInsert => {
+                            self.buffer.insert((timestamp, pk), row.clone());
```

Comment:
Well, it doesn't seem like buffer is a cache, and we seem to assume we can buffer the entire range from the temporal filter condition to now. Anw, this may OOM. 

It may be a better idea to simply implement a prefetching buffer than prefetches the next K rows where K is the average number of rows read from from storage in the last N epochs.

Anw, this is probably a step for next PR.

Reply:
> Well, it doesn't seem like buffer is a cache, and we seem to assume we can buffer the entire range satisfying the temporal filter condition. Anw, this may OOM.
> 
> It may be a better idea to simply implement a prefetching buffer than prefetches the next K rows where K is the average number of rows read from from storage in the last N epochs.
> 
> Anw, this is probably a step for next PR.

yes, if we create an independent `temporal filter`, we can do as what `dynamic filter` does. We can also use `dynamic filter` and `now` to implement temporal filter directly.